Add _spacers.less to generate pad/mar utility classes

benjaminapetersen opened the pull request:
Adds `_spacers.jess` to generate margin & padding utility classes

Discussion generated originally from [this PR](https://github.com/openshift/origin/pull/8180)

CSS output will look something like this:

```css
.pad-none{padding:0px}
 .mar-none{margin:0px}
 .pad-top-none{padding-top:0px!important}
 .mar-top-none{margin-top:0px!important}
 .pad-right-none{padding-right:0px!important}
 .mar-right-none{margin-right:0px!important}
 .pad-bottom-none{padding-bottom:0px!important}
 .mar-bottom-none{margin-bottom:0px!important}
 .pad-left-none{padding-left:0px!important}
 .mar-left-none{margin-left:0px!important}
 .pad-sm{padding:2px 4px}
 .mar-sm{margin:2px 4px}
 .pad-top-sm{padding-top:4px}
 .mar-top-sm{margin-top:4px}
 .pad-right-sm{padding-right:4px}
 .mar-right-sm{margin-right:4px}
 .pad-bottom-sm{padding-bottom:4px}
 .mar-bottom-sm{margin-bottom:4px}
 .pad-left-sm{padding-left:4px}
 .mar-left-sm{margin-left:4px}
 .pad-md{padding:4px 8px}
 .mar-md{margin:4px 8px}
 .pad-top-md{padding-top:8px}
 .mar-top-md{margin-top:8px}
 .pad-right-md{padding-right:8px}
 .mar-right-md{margin-right:8px}
 .pad-bottom-md{padding-bottom:8px}
 .mar-bottom-md{margin-bottom:8px}
 .pad-left-md{padding-left:8px}
 .mar-left-md{margin-left:8px}
 .pad-lg{padding:6px 12px}
 .mar-lg{margin:6px 12px}
 .pad-top-lg{padding-top:12px}
 .mar-top-lg{margin-top:12px}
 .pad-right-lg{padding-right:12px}
 .mar-right-lg{margin-right:12px}
 .pad-bottom-lg{padding-bottom:12px}
 .mar-bottom-lg{margin-bottom:12px}
 .pad-left-lg{padding-left:12px}
 .mar-left-lg{margin-left:12px}
 .pad-xl{padding:8px 16px}
 .mar-xl{margin:8px 16px}
 .pad-top-xl{padding-top:16px}
 .mar-top-xl{margin-top:16px}
 .pad-right-xl{padding-right:16px}
 .mar-right-xl{margin-right:16px}
 .pad-bottom-xl{padding-bottom:16px}
 .mar-bottom-xl{margin-bottom:16px}
 .pad-left-xl{padding-left:16px}
 .mar-left-xl{margin-left:16px}
 .pad-xxl{padding:16px 32px}
 .mar-xxl{margin:16px 32px}
 .pad-top-xxl{padding-top:32px}
 .mar-top-xxl{margin-top:32px}
 .pad-right-xxl{padding-right:32px}
 .mar-right-xxl{margin-right:32px}
 .pad-bottom-xxl{padding-bottom:32px}
 .mar-bottom-xxl{margin-bottom:32px}
 .pad-left-xxl{padding-left:32px}
 .mar-left-xxl{margin-left:32px}
```
Usage example:
```html
<div row reverse class="pad-md pad-left-none">
  <div flex class="pad-lg">Foo</div>
</div>
```


@jwforres @spadgett @rhamilto @sg00dwin breaking this out as a separate PR so we can discuss.  Items on the table:

- [x] Yay, will provide easy way to eliminate inline styles (cough, @rhamilto)
- [x] Is a simple consistent pattern that can be updated/extended by adjusting the `@sizes` or `@sides` list
- [x] Is consistent-ish with where Alpha bootstrap is going, though different naming convention & provides more options.  I expect Bootstrap will iterate on their naming 
- [x] is probably the same (or very close) as to what will land in `layout.attrs` in terms of naming, etc

Issues to address:
- [x] We use 5px, 10px throughout app, though have a variety of overrides/differences sprinkled around.  Would be good to normalize & ensure we set the `@sizes @sides` with the px vals we want to use.
- [x] Isn't exactly semantic, but is consistent w/bootstrap * layout.attrs methods of composable utilities

I think that sums up the discussion points to address.

jwforres commented:
fine with the general direction of the PR, i will leave this up to @rhamilto and @sg00dwin to review the changes

benjaminapetersen commented:
- [ ] @rhamilton hates `0px`, fix that :smile:

benjaminapetersen commented:
@rhamilto @sg00dwin, pinging to see if either of you have more thoughts on this

spadgett commented on assets/app/styles/_spacers.less:
```diff
@@ -0,0 +1,36 @@
+//  .mar-sm, .mar-md, .mar-lg, etc
+//  .pad-left-sm, .mar-left-sm, etc
+//
+@sizes: none 0px e('!important'), sm 4px e(''), md 8px e(''), lg 12px e(''), xl 16px e(''), xxl 32px e('');
```
Can we make one consistent with the `gutter-top` length?

rhamilto replied:
Should the size values be variables so they can easily be overridden?

benjaminapetersen replied:
Prob should use `5px` at one size for consistency, agree.

The `@sizes` variable is the less equivalent of an array.  I was thinking we would override the whole array if desired, but good Q.  That could live in our `_variables.less` instead, though im not sure if its best to just keep it all in the same place so it doesnt get confused with other stuff...

benjaminapetersen replied:
Btw, bootstrap's version is based on `rem`, this isn't.  At present I'm not super accustomed to generating layout based on root em value. Thoughts on that welcome (but may be good not to adopt the BS naming convention then, as they probably will be quite different values).

As a pattern, `2,4,8,12,16,32` work well and are easy to generate balanced top,bottom, left, right rules.  To match `.gutter-top` we could go with `5,10,15,20,25,30`, but it would be interesting to know what px vals would actually look best.  


Updated the vars to look like this:

```less
@nil: none 0 e('!important');
// @auto 0 auto; is built into the fn below
@xs: xs 3px e('');      // 2px
@sm: sm 5px e('');      // 4px
@md: md 10px e('');     // 8px
@lg: lg 15px e('');     // 12px
@xl: xl 20px e('');     // 16px
@xxl: xxl 30px e('');   // 32px

@sizes: @nil, @xs, @sm, @md, @lg, @xl, @xxl;
@sides: top, right, bottom, left;
```

Which will generate this:
```css
.pad-auto-none {padding: 0 auto; }
.mar-auto-none {margin: 0 auto; }
.pad-none {padding: 0 0; }
.mar-none {margin: 0 0; }
.pad-top-none {padding-top: 0 !important; }
.mar-top-none {margin-top: 0 !important; }
.pad-right-none {padding-right: 0 !important; }
.mar-right-none {margin-right: 0 !important; }
.pad-bottom-none {padding-bottom: 0 !important; }
.mar-bottom-none {margin-bottom: 0 !important; }
.pad-left-none {padding-left: 0 !important; }
.mar-left-none {margin-left: 0 !important; }
.pad-auto-xs {padding: 0 auto; }
.mar-auto-xs {margin: 0 auto; }
.pad-xs {padding: 3px 4.5px; }
.mar-xs {margin: 3px 4.5px; }
.pad-top-xs {padding-top: 3px ; }
.mar-top-xs {margin-top: 3px ; }
.pad-right-xs {padding-right: 3px ; }
.mar-right-xs {margin-right: 3px ; }
.pad-bottom-xs {padding-bottom: 3px ; }
.mar-bottom-xs {margin-bottom: 3px ; }
.pad-left-xs {padding-left: 3px ; }
.mar-left-xs {margin-left: 3px ; }
.pad-auto-sm {padding: 0 auto; }
.mar-auto-sm {margin: 0 auto; }
.pad-sm {padding: 5px 7.5px; }
.mar-sm {margin: 5px 7.5px; }
.pad-top-sm {padding-top: 5px ; }
.mar-top-sm {margin-top: 5px ; }
.pad-right-sm {padding-right: 5px ; }
.mar-right-sm {margin-right: 5px ; }
.pad-bottom-sm {padding-bottom: 5px ; }
.mar-bottom-sm {margin-bottom: 5px ; }
.pad-left-sm {padding-left: 5px ; }
.mar-left-sm {margin-left: 5px ; }
.pad-auto-md {padding: 0 auto; }
.mar-auto-md {margin: 0 auto; }
.pad-md {padding: 10px 15px; }
.mar-md {margin: 10px 15px; }
.pad-top-md {padding-top: 10px ; }
.mar-top-md {margin-top: 10px ; }
.pad-right-md {padding-right: 10px ; }
.mar-right-md {margin-right: 10px ; }
.pad-bottom-md {padding-bottom: 10px ; }
.mar-bottom-md {margin-bottom: 10px ; }
.pad-left-md {padding-left: 10px ; }
.mar-left-md {margin-left: 10px ; }
.pad-auto-lg {padding: 0 auto; }
.mar-auto-lg {margin: 0 auto; }
.pad-lg {padding: 15px 22.5px; }
.mar-lg {margin: 15px 22.5px; }
.pad-top-lg {padding-top: 15px ; }
.mar-top-lg {margin-top: 15px ; }
.pad-right-lg {padding-right: 15px ; }
.mar-right-lg {margin-right: 15px ; }
.pad-bottom-lg {padding-bottom: 15px ; }
.mar-bottom-lg {margin-bottom: 15px ; }
.pad-left-lg {padding-left: 15px ; }
.mar-left-lg {margin-left: 15px ; }
.pad-auto-xl {padding: 0 auto; }
.mar-auto-xl {margin: 0 auto; }
.pad-xl {padding: 20px 30px; }
.mar-xl {margin: 20px 30px; }
.pad-top-xl {padding-top: 20px ; }
.mar-top-xl {margin-top: 20px ; }
.pad-right-xl {padding-right: 20px ; }
.mar-right-xl {margin-right: 20px ; }
.pad-bottom-xl {padding-bottom: 20px ; }
.mar-bottom-xl {margin-bottom: 20px ; }
.pad-left-xl {padding-left: 20px ; }
.mar-left-xl {margin-left: 20px ; }
.pad-auto-xxl {padding: 0 auto; }
.mar-auto-xxl {margin: 0 auto; }
.pad-xxl {padding: 30px 45px; }
.mar-xxl {margin: 30px 45px; }
.pad-top-xxl {padding-top: 30px ; }
.mar-top-xxl {margin-top: 30px ; }
.pad-right-xxl {padding-right: 30px ; }
.mar-right-xxl {margin-right: 30px ; }
.pad-bottom-xxl {padding-bottom: 30px ; }
.mar-bottom-xxl {margin-bottom: 30px ; }
.pad-left-xxl {padding-left: 30px ; }
.mar-left-xxl {margin-left: 30px ; }
```

spadgett replied:
We tend to use multiples of 5px today. Not that we have to stick to that.

A quick check with some fuzzy, probably incorrect regex.

```
assets❯ ag --less --html '(padding|margin).*:.* 5px' | wc -l                                                                                                                                                                                     origin/git/debug-terminal-1.3
      55
assets❯ ag --less --html '(padding|margin).*:.* 10px' | wc -l                                                                                                                                                                                    origin/git/debug-terminal-1.3
      74
assets❯ ag --less --html '(padding|margin).*:.* 15px' | wc -l                                                                                                                                                                                    origin/git/debug-terminal-1.3
      17
assets❯ ag --less --html '(padding|margin).*:.* 20px' | wc -l                                                                                                                                                                                    origin/git/debug-terminal-1.3
      30
assets❯ ag --less --html '(padding|margin).*:.* 30px' | wc -l                                                                                                                                                                                    origin/git/debug-terminal-1.3
      13
assets❯ ag --less --html '(padding|margin).*:.* 2px' | wc -l                                                                                                                                                                                     origin/git/debug-terminal-1.3
       8
assets❯ ag --less --html '(padding|margin).*:.* 3px' | wc -l                                                                                                                                                                                     origin/git/debug-terminal-1.3
      13
assets❯ ag --less --html '(padding|margin).*:.* 8px' | wc -l                                                                                                                                                                                     origin/git/debug-terminal-1.3
       4
assets❯ ag --less --html '(padding|margin).*:.* 12px' | wc -l                                                                                                                                                                                    origin/git/debug-terminal-1.3
       4
```

benjaminapetersen replied:
Yeah i definitely wanted to cover the 5,10,15,20... and also be flexible enough for the deviants like 2,3,8,12.  Half steps are always helpful, but `xs,sm,md,lg,xl,xxl` seems like it should be enough for what we need.  Perhaps there is a better naming convention to give us a few more options?

spadgett replied:
xs - 3, sm - 5, md - 10, lg - 20, xl - 30, xxl - 60 ?

benjaminapetersen replied:
I'm good with that... if @sg00dwin & @rhamilto agree, I can make the change.  I think if we have a consensus on the pattern we can update & run with it.

spadgett commented on assets/app/styles/_spacers.less:
```diff
@@ -0,0 +1,51 @@
+@md: md 10px e('');     // 8px
+@lg: lg 15px e('');     // 12px
+@xl: xl 20px e('');     // 16px
+@xxl: xxl 30px e('');   // 32px
```
Looks like the comments no longer match the values.

benjaminapetersen replied:
Yup, left my original vals in comments as we discussed in case I wanted/needed to return to them.  I'll squash this & remove them.

benjaminapetersen commented:
Squashed, removed comments.  Should be good to go if no objections.  We can always use them a bit & tweak/update the vars if they don't quite meet the needs.

sg00dwin commented:
@benjaminapetersen what's the reasoning behind the difference of values on the on the non targeted side classes  ```.pad-md{padding:4px 8px} .mar-md{margin:4px 8px}``` . bootstrap sets 'all' equally. 

Personally I think it adds complexity of having to know separate values for each. If we have utility spacing, with one of the desired outcomes is the ability to read the classes and quickly understand these rules, then if someone wants to achieve non equal space for left/right than top/bottom, wouldn't it make sense to have ```class="pad-lg pad-left-md pad-right-md"```. Thus I know they are different.

benjaminapetersen commented:
@sg00dwin because rarely do we want squares, usually we deal with rectangles.  Top & bottom padding are not terribly useful if they match the sides.  The (Bootstrap 4 alpha)[http://v4-alpha.getbootstrap.com/components/utilities/#spacing] classes are doing the same thing via something like: ` ($spacer-x * 1.5)`.

benjaminapetersen commented:
Though, it would be easy enough to generate a few more classes w/consistent naming `sq-sm, sq-md, sq-lg, etc.` if we wanted classes for making squares.

sg00dwin commented:
They have classes that target just the 'x' axis and 'y' axis. But their 'all sides' rules ```p-a-1``` have the same value. https://github.com/twbs/bootstrap/blob/v4-dev/dist/css/bootstrap.css#L6171

```
.p-a-1 {
  padding: 1rem 1rem !important;
}

.p-t-1 {
  padding-top: 1rem !important;
}

.p-r-1 {
  padding-right: 1rem !important;
}

.p-b-1 {
  padding-bottom: 1rem !important;
}

.p-l-1 {
  padding-left: 1rem !important;
}

.p-x-1 {
  padding-right: 1rem !important;
  padding-left: 1rem !important;
}

.p-y-1 {
  padding-top: 1rem !important;
  padding-bottom: 1rem !important;
}
```

benjaminapetersen commented:
Intersting. They also put `!important;` on everything.  I considered that but held off, however it is probably appropriate to use.  If you are using `pad-top-sm` or something like it, you definitely want it to win and override other styles applied.

benjaminapetersen commented:
So if `pad-md` = `padding: 5px 10px;` and `pad-all-md` = `padding: 5px;` is that sensible?  I'm sure this can be updated to an easy/consistent pattern.

sg00dwin commented:
if we want to apply different spacing on the horiz/vert axis why not have comparable 'x' or 'y' axis classes then use ```class="pad-y-lg pad-x-md``` to vary spacing...

And back to my original comment "... one of the desired outcomes is the ability to read the classes and quickly understand these rules" isn't the above more understandable.

benjaminapetersen commented:
The `.pad-x-md` and `.pad-y-md` pattern makes sense for "sides" and "top-bottom" i think, since there is no good word for "top-bottom".

I'm completely up for a follow-up PR & making any helpful additions/modifications.  Lets test them a bit and see how they work.

spadgett commented:
> Lets test them a bit and see how they work.

+1, I think we should try these for a while before adding too many more styles.

benjaminapetersen commented:
I've used basically this pattern for scaffolding in some of my `/demo` directories for angular plugins (i think the extension-registry, element-query, etc).  They have worked well for snapping quick widgets together, but building some real stuff would be a good test.
